tests/test_timer.cpp

```diff
@@ -32,165 +33,280 @@ using namespace testing;
 
 namespace
 {
-struct Timer : public testing::Test
+struct MockTimerSyncFuncs : public mpu::TimerSyncFuncs
 {
-    std::atomic_bool timedout = false;
+    using TimerSyncFuncs::TimerSyncFuncs;
+
+    MOCK_CONST_METHOD1(notify_all, void(std::condition_variable&));
+    MOCK_CONST_METHOD2(wait, void(std::condition_variable&, std::unique_lock<std::mutex>&));
+    MOCK_CONST_METHOD3(wait_for, std::cv_status(std::condition_variable&, std::unique_lock<std::mutex>&,
+                                                const std::chrono::duration<int, std::milli>&));
+
+    MP_MOCK_SINGLETON_BOILERPLATE(MockTimerSyncFuncs, TimerSyncFuncs);
+};
+
+struct TestTimer : public testing::Test
+{
+    TestTimer()
+        : t{default_timeout, [this] {
+                {
+                    std::lock_guard<std::mutex> test_lock{cv_m};
+                    ++timeout_count;
+                }
+                cv.notify_all();
+            }}
+    {
+    }
+
+    auto mock_wait(std::unique_lock<std::mutex>& timer_lock, std::atomic_bool& target_state)
+    {
+        std::unique_lock<std::mutex> test_lock(cv_m);
+
+        // Unlock the Timer's internal lock
+        timer_lock.unlock();
+
+        // Notify the test thread that the timer is running
+        target_state.store(true);
+        cv.notify_all();
+
+        // Wait for notification from main test thread
+        cv.wait(test_lock, [this] { return test_notify_called.load(); });
+        test_notify_called.store(false);
+
+        // Re-lock the Timer's internal lock before returning
+        timer_lock.lock();
+    }
+
+    auto make_mock_wait_for(std::cv_status timeout_status)
+    {
+        return [this, timeout_status](std::unique_lock<std::mutex>& timer_lock) {
+            mock_wait(timer_lock, timer_running);
+            return timeout_status;
+        };
+    }
+
+    void wait_for(const std::atomic_bool& target_state)
+    {
+        std::unique_lock<std::mutex> test_lock(cv_m);
+        cv.wait(test_lock, [&target_state] { return target_state.load(); });
+    }
+
+    void notify_test_wait()
+    {
+        test_notify_called.store(true);
```

Comment:
It would be nice to lock inside this function, [because](https://en.cppreference.com/w/cpp/thread/condition_variable):

> Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread. 

Then the caller doesn't have to. I understand you sometimes need the lock to endure, so it can be returned. Here's [what I mean ](https://github.com/canonical/multipass/commit/3f81125a3306af7653e4728684d0f6b709560b29) (in a commit because GH suggestions not very suitable for different bits of code).

Reply:
Right, I started doing that, but then thought about the lock being needed afterwards in some cases.  I obviously didn't think to *return* the lock :wink:  Good idea, and I'll add this since this is really what I wanted to do.  Thanks!